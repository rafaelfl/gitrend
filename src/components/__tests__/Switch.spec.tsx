import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Switch } from '..';

describe('Switch component tests', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    test('it should match snapshot', () => {
        const checkedSwitch = render(<Switch checked={true} onChange={() => false} />);
        const uncheckedSwitch = render(<Switch checked={false} onChange={() => false} />);

        expect(checkedSwitch.container).toMatchSnapshot();
        expect(uncheckedSwitch.container).toMatchSnapshot();
    });

    test('it should change the checked state on click event', () => {
        let isChecked = false;

        const onChange = jest.fn(() => (isChecked = !isChecked));

        const { rerender, container } = render(<Switch checked={isChecked} onChange={onChange} />);

        const switchContainer = container;
        const checkbox = switchContainer.querySelector('input[type="checkbox"]') as Element;
        const spanCheckbox = switchContainer.querySelector('span');

        expect(switchContainer).toBeInTheDocument();
        expect(checkbox).toBeInTheDocument();
        expect(spanCheckbox).toBeInTheDocument();

        userEvent.click(checkbox);

        expect(isChecked).toEqual(true);
        expect(onChange).toHaveBeenCalled();

        rerender(<Switch checked={isChecked} onChange={onChange} />);

        expect(checkbox).toBeChecked();
    });

    test('it should not change the checked state on click event because it is disabled', () => {
        let isChecked = false;

        const onChange = jest.fn(() => (isChecked = !isChecked));

        const { rerender, container } = render(<Switch checked={isChecked} disabled onChange={onChange} />);

        const switchContainer = container;
        const checkbox = switchContainer.querySelector('input[type="checkbox"]') as Element;
        const spanCheckbox = switchContainer.querySelector('span');

        expect(switchContainer).toBeInTheDocument();
        expect(checkbox).toBeInTheDocument();
        expect(spanCheckbox).toBeInTheDocument();

        userEvent.click(checkbox);

        expect(isChecked).toEqual(false);
        expect(onChange).not.toHaveBeenCalled();

        rerender(<Switch checked={isChecked} onChange={onChange} />);

        expect(checkbox).not.toBeChecked();
    });

    test('it should change the checked state on enter keydown', () => {
        let isChecked = false;

        const onChange = jest.fn(() => {
            isChecked = !isChecked;
        });

        const { rerender, container } = render(<Switch checked={isChecked} onChange={onChange} />);

        const switchContainer = container;
        const checkbox = switchContainer.querySelector('input[type="checkbox"]') as Element;
        const spanCheckbox = switchContainer.querySelector('span');

        expect(switchContainer).toBeInTheDocument();
        expect(checkbox).toBeInTheDocument();
        expect(spanCheckbox).toBeInTheDocument();

        fireEvent.keyDown(checkbox, { key: 'A' });
        fireEvent.keyDown(checkbox, { key: 'Enter' });

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(isChecked).toEqual(true);

        rerender(<Switch checked={isChecked} onChange={onChange} />);

        expect(checkbox).toBeChecked();
    });
});
