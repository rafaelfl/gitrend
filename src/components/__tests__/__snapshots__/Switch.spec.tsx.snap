// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component tests it should match snapshot 1`] = `
<div>
  <label
    aria-label="Switch button"
    class="sc-jRQBWg krnHGF"
    data-testid="@Switch/container"
    tabindex="0"
  >
    <input
      checked=""
      data-testid="@Switch/checkbox"
      type="checkbox"
    />
    <span />
  </label>
</div>
`;

exports[`Switch component tests it should match snapshot 2`] = `
<div>
  <label
    aria-label="Switch button"
    class="sc-jRQBWg krnHGF"
    data-testid="@Switch/container"
    tabindex="0"
  >
    <input
      data-testid="@Switch/checkbox"
      type="checkbox"
    />
    <span />
  </label>
</div>
`;
